### ISSUE TITLE
[BUG] dask_cudf.read_parquet fails on partitioned data

### ISSUE THREAD

**Issue by randerzander** [rapidsai/cudf#2829]
Possibly related to #2683, using dask_cudf.read_parquet on the `test_pandas` dir fails:
```
import pandas as pd
import dask.dataframe as ddf
import dask_cudf

df = pd.DataFrame()
df['year'] = [2018, 2019, 2019, 2019]
df['month'] = [1, 2, 3, 3]
df['day'] = [1, 1, 1, 2]
df['data'] = [0, 0, 0, 0]

ddf = ddf.from_pandas(df, npartitions=1)
ddf.to_parquet('test_pandas', partition_on=['year', 'month', 'day'])

dask_cudf.read_parquet('test_pandas').compute()
```

Result:
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-12-5ece4ef70323> in <module>
----> 1 dask_cudf.read_parquet('test_pandas').compute()

/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py in compute(self, **kwargs)
    173         dask.base.compute
    174         """
--> 175         (result,) = compute(self, traverse=False, **kwargs)
    176         return result
    177 

/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    444     keys = [x.__dask_keys__() for x in collections]
    445     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 446     results = schedule(dsk, keys, **kwargs)
    447     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    448 

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in get_sync(dsk, keys, **kwargs)
    530     """
    531     kwargs.pop("num_workers", None)  # if num_workers present, remove it
--> 532     return get_async(apply_sync, 1, dsk, keys, **kwargs)
    533 
    534 

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    474             # Seed initial tasks into the thread pool
    475             while state["ready"] and len(state["running"]) < num_workers:
--> 476                 fire_task()
    477 
    478             # Main loop, wait on tasks to finish, insert new ones

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in fire_task()
    469                         pack_exception,
    470                     ),
--> 471                     callback=queue.put,
    472                 )
    473 

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in apply_sync(func, args, kwds, callback)
    519 def apply_sync(func, args=(), kwds={}, callback=None):
    520     """ A naive synchronous version of apply_async """
--> 521     res = func(*args, **kwds)
    522     if callback is not None:
    523         callback(res)

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    236         failed = False
    237     except BaseException as e:
--> 238         result = pack_exception(e, dumps)
    239         failed = True
    240     return key, result, failed

/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    231     try:
    232         task, data = loads(task_info)
--> 233         result = _execute_task(task, data)
    234         id = get_id()
    235         result = dumps((result, id))

/conda/envs/rapids/lib/python3.7/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py in read_parquet_part(func, fs, meta, part, columns, index, kwargs)
    304     columns = columns or []
    305     index = index or []
--> 306     return df[[c for c in columns if c not in index]]
    307 
    308 

/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/dataframe.py in __getitem__(self, arg)
    354                     return df
    355                 for col in arg:
--> 356                     df[col] = self[col]
    357             return df
    358         elif isinstance(arg, DataFrame):

/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/dataframe.py in __getitem__(self, arg)
    323             return self.columns._get_column_major(self, arg)
    324         if is_scalar(arg) or isinstance(arg, tuple):
--> 325             s = self._cols[arg]
    326             s.name = arg
    327             s.index = self.index

KeyError: 'day'
```

**Comment by kkraus14** [comment]
I'm guessing this is because `cudf.read_parquet` doesn't handle directory partitioning while `pandas` / `pyarrow` do. @rjzamora How does Dask end up calling `cudf.read_parquet` in this situation?

**Comment by rjzamora** [comment]
I'm looking into this now - You are correct that dask assumes pyarrow and fastparquet can handle a `partitions` kwarg to take care of this.  In our `dask_cudf` `read_partition` implementation, we have access to a `partitions` argument of type `pyarrow.parquet.ParquetPartitions`, but `cudf.read_parquet` obviously cannot do anything with this information at the moment.

The best long term solution is to handle partitioning in `cudf.read_parquet`.  However, I can certainly add the necessary logic to our `read_partition` implementation in the meantime.

**Comment by kkraus14** [comment]
Will open an issue for cuIO, thanks!

**Comment by rjzamora** [comment]
> I think we should just handle in the cudf.read_parquet code as there's partitioning arguments in the writers at the Pandas level that could potentially break things as well.

> Will open an issue for cuIO, thanks!

Sounds good @kkraus - BTW, I just confirmed that the problem is resolved if you add the following logic after the call to `cudf.read_parquet` in `read_partition` (`cudf/python/cudf/cudf/io/parquet.py`):

```python
if len(piece.partition_keys) > 0:
    # Partitioning logic copied from
    # (apache/arrow/python/pyarrow/parquet.py)
    if partitions is None:
        raise ValueError('Must pass partition sets')
    import numpy as np
    import pyarrow as pa
    for i, (name, index2) in enumerate(piece.partition_keys):
        indices = np.array([index2], dtype='i4').repeat(len(df))
        dictionary = partitions.levels[i].dictionary
        arr = pa.DictionaryArray.from_arrays(indices, dictionary).to_pandas()
        df[name] = cudf.Series(arr).astype(arr.dtype._categories.dtype)
```

Similar logic could be moved into `cudf.read_parquet` if we want that function to accept `pyarrow`-based partition information.

**Comment by kkraus14** [comment]
Issue opened at #2845

**Comment by kkraus14** [comment]
@rjzamora if you want to throw a workaround in within `dask-cudf` for now I think that may be good. I'm not sure how difficult this would be to handle properly in the C++ / CUDA level.

**Comment by randerzander** [comment]
This was closed by #2850
